### PR TITLE
start of cleaning

### PR DESCRIPTION
**Pull request by WillHK**

### CODE REVIEW

**Review by hughjafro — COMMENTED** [review]
Module 1:

- Start a clean notebook.
  - Good use of comments
- Do train/validate/test split with the Tanzania Waterpumps data.
  - Stretch: Explore different train_size and test_size
- Begin to explore and clean the data.
  - Import has the correct shape
  - Good methodical approach to analyze each feature
  - I would be interested to know your thought process in creating the function return_mean_if_zero
  - Good start to analyze for cardinality
  - I like your implementations of functions. Make sure they are not overkill, in terms of having a one-time use or can be solved with a builtin.
- Select different numeric and categorical features.
  - Your cat features selections are a good choice
- Do one-hot encoding. (Remember it may not work with high cardinality categoricals.)
- Scale features.
- Use scikit-learn for logistic regression.
  - Check
  - Does using `solver='liblinear' in your LR model make a difference in score?
- Get your validation accuracy score.
  - nice score
- Get and plot your coefficients.
  - How you might address and impute missing/zero values have an effect on model scores, try different ways
- Submit your predictions to our Kaggle competition.
- Commit your notebook to your fork of the GitHub repo.

**Review by hughjafro — COMMENTED** [review]
**Module2 DecisionTree:**
- I might have missed it, but did not see the use of DecisionTree. You may not have committed that notebook yet.

- Try to explore GradientBoosting and RandomForest. GB should provide a better score than RF, but only if your parameter tuning is carefully done.
  - GB is more prone to overfitting (higher variance) and generally takes longer because the trees are built one at a time/sequentially. Takes a bit of work to tune properly, but once you do the results are generally better than RF.
  - RF is good for multiclass problems and is easier to tune than GB. When there are categoricals, RF will tend to bias those cats with more levels (partial permutation methods help with this)

The following is a good read: [RF or GB?](https://www.quora.com/When-would-one-use-Random-Forests-over-Gradient-Boosted-Machines-GBMs)

**Comment by hughjafro** [comment]
### Module 03 Random Forest:

**Overall:** TOP 3!!! Hang On!
**Feedback:** Comment your code. 

_Do more exploratory data analysis, data cleaning, feature engineering, and feature selection._
- What happens if you split "controversial" features into 2/3/4 parts?

_Try a Random Forest Classifier._
- Check

_Try Ordinal Encoding._
- Check

Try changing your Xgboost parameters: subsample, max_depth, scale_pos_weight, reg_alpha.

**Submit new predictions.**

**Comment by hughjafro** [comment]
### Module 04 Confusion Matrix:

**Overall:** 🚀🚀🚀🚀🚀Great Job 🚀🚀🚀🚀🚀
**Feedback:** Your code is always well organized and easy to read.

_Get and interpret the confusion matrix for classification models_
- At least take a stab at the confusion matrix
- What does the confusion matrix tell you?
- What is the significance of the precision, recall, f-1 scores?

_Kaggle Challenge_
- Your perseverance and hard work paid 💰 off.
- Very nice and clear kernel.
- What is the most important thing that you learned in this kaggle challenge? Write a blog about it.
- Did you have problems with overfitting or underfitting? If so, how was it handled?
- Ask yourself, what you would do differently on your next Kaggle challenge? In approach and execution.
- Don't give up on xgboost 🚀. This is a kaggle mainstay.

**Congrats Champ** 🥊
